oceanmodes/baroclinic.py

```diff
@@ -165,7 +168,240 @@ def _neutral_modes_from_N2_profile_raw(z, N2, f0, depth=None, **kwargs):
+
+def _instability_analysis_from_N2_profile_raw(z, N2, f0, beta, Nx, Ny, dx, dy, ubar, vbar, etax, etay, sort='LI', num=4, depth=None, **kwargs):
+    """Calculate baroclinic unstable modes from a profile of buoyancy frequency.
+    
+        Solves the eigenvalue problem
+
+            \left ( \frac{\partial}{\partial t} + U \cdot \nabla \right ) \left ( \nabla^2 + \frac{d}{dz} \frac{f_0^2}{N^2} \frac{d \psi}{d z} \right )
+            + u \cdot \nabla Q
+            = 0
+
+        With the boundary conditions
+
+            \left ( \frac{\partial}{\partial t} + U \cdot \nabla \right ) f_0 \frac{\partial \psi}{\partial z}
+            + u \cdot \nabla \left ( B + N^2 \eta \right )
+            = 0
+
+        at the top and bottom.
+    
+        Parameters
+        ----------
+        z : array_like
+            The depths at which N2 is given. Starts shallow, increases
+            positive downward.
+        N2 : array_like
+            The squared buoyancy frequency (units s^-2). Points below topography
+            should be masked or nan.
+        f0 : float
+            Coriolis parameter.
+        Q : array_like
+            The background QGPV (\beta y + \left ( \nabla^2 + \frac{d}{dz} \frac{f_0^2}{N^2} \frac{d \Psi}{d z} \right ))
+            and Psi is the streamfunction of the background velocity.
+        eta : array_like
+            The topographic change at the top and bottom.
+        B : array_like
+            The background buoyancy
+        depth : float, optional
+            Total water column depth. If missing will be inferred from z.
+        kwargs : optional
+            Additional parameters to pass to scipy.sparse.linalg.eigs for
+            eigenvalue computation
+    
+        Returns
+        -------
+        zf : array_like
+            The depths at which phi is defined. Different from z.
+        psi : array_like
+            vertical modes
+        omega: array_like
+            growth rate
+        k, l : array_like
+            inverse horizontal wavelengths
+    """
+    nz = len(z)
+
+    ### vertical discretization ###
+
+    # ~~~~~ zf[0]==0, phi[0] ~~~~
+    #
+    # ----- zc[0], N2[0] --------
+    #
+    # ----- zf[1], phi[1] -------
+    # ...
+    # ---- zc[nz-1], N2[nz-1] ---
+    #
+    # ~~~~ zf[nz], phi[nz] ~~~~~~
+
+    # just for notation's sake
+    # (user shouldn't worry about discretization)
+    zc = z
+    dzc = np.hstack(np.diff(zc))
+    # make sure z is increasing
+    if not np.all(dzc > 0):
+        raise ValueError('z should be monotonically increasing')
+    if depth is None:
+        depth = z[-1] + dzc[-1]/2
+    else:
+        if depth <= z[-1]:
+            raise ValueError('depth should not be less than maximum z')
+
+    dztop = zc[0]
+    dzbot = depth - zc[-1]
+
+    # put the phi points right between the N2 points
+    zf = np.hstack([0, 0.5*(zc[1:]+zc[:-1]), depth ])
+    dzf = np.diff(zf)
```

Comment:
A lot of this code seems just copied directly from neutral_neutral_modes_from_N2_profile. That means it should be moved into a function of its own.